File: Lab06Cpp/Sensor.cpp
#include "Sensor.hpp"

Sensor::Sensor(const std::string& id,
               const std::string& name,
               const std::string& unit)
    : m_id(id)
    , m_name(name)
    , m_unit(unit)
    , m_lastValue(0.0f)
    , m_status("OK")
{
}

bool Sensor::CheckStatus() const
{
    std::cout << "Checking status of sensor '" << m_name
              << "': status = " << m_status << std::endl;
    return m_status == "OK";
}

void Sensor::Calibrate()
{
    std::cout << "Base calibration for sensor '" << m_name << "'" << std::endl;
}

std::string Sensor::Describe() const
{
    return "Sensor[" + m_id + "] '" + m_name + "' (" + m_unit + ")";
}

void Sensor::PrintWithHeader() const
{
    std::cout << "[Sensor info] " << Describe() << std::endl;
}

std::string Sensor::TypeName() const
{
    // Невиртуальная функция: при вызове через Sensor* будет использоваться
    // именно эта реализация, даже если в потомке есть функция с тем же именем.
    return "BaseSensor";
}

Sensor& Sensor::SetStatus(const std::string& status)
{
    this->m_status = status; // Явное использование this.
    return *this;
}

File: Lab06Cs/Lab06Monitoring/Sensor.cs
using System;

namespace Lab06Monitoring
{
    /// <summary>
    /// Абстрактный базовый класс для любых датчиков.
    ///
    /// ЛР-6:
    ///  * демонстрируем protected-свойства (protected set);
    ///  * вводим виртуальные методы (ReadValue, Describe, Clone);
    ///  * демонстрируем вызов виртуального метода через невиртуальный (PrintWithHeader);
    ///  * показываем разницу между виртуальной и невиртуальной функцией (Describe vs TypeName);
    ///  * реализуем поверхностное/глубокое клонирование в связке с Room.
    /// </summary>
    public abstract class Sensor : ICloneable
    {
        public string Id { get; }
        public string Name { get; }
        public string Unit { get; }

        public float LastValue { get; protected set; }
        public string Status { get; protected set; } = "OK";

        protected Sensor(string id, string name, string unit)
        {
            Id = id;
            Name = name;
            Unit = unit;
        }

        /// <summary>
        /// Виртуальное чтение значения.
        /// </summary>
        public abstract float ReadValue();

        /// <summary>
        /// Виртуальная проверка состояния.
        /// </summary>
        public virtual bool CheckStatus()
        {
            Console.WriteLine($"Checking status of sensor '{Name}': {Status}");
            return Status == "OK";
        }

        /// <summary>
        /// Виртуальная калибровка.
        /// </summary>
        public virtual void Calibrate()
        {
            Logger.Instance.Log($"Base calibration for sensor {Id}");
        }

        /// <summary>
        /// Виртуальное описание датчика.
        /// </summary>
        public virtual string Describe()
        {
            return $"Sensor[{Id}] '{Name}' ({Unit})";
        }

        /// <summary>
        /// Невиртуальный метод, который вызывает виртуальный Describe().
        /// </summary>
        public void PrintWithHeader()
        {
            Console.WriteLine("[Sensor info] " + Describe());
        }

        /// <summary>
        /// Невиртуальный метод. В производных классах объявим метод с тем же
        /// именем и сигнатурой, чтобы показать разницу при вызове через базовый
        /// и производный тип.
        /// </summary>
        public string TypeName()
        {
            return "BaseSensor";
        }

        /// <summary>
        /// Поддержка поверхностного клонирования через ICloneable.
        /// Конкретные классы переопределяют Clone(), возвращая свой тип.
        /// </summary>
        public abstract object Clone();

        /// <summary>
        /// Метод, демонстрирующий использование this (возвращаем сам объект).
        /// </summary>
        public Sensor SetStatus(string status)
        {
            this.Status = status; // явное использование this
            return this;
        }
    }
}
